### PR TITLE
Tile sheet with different sized tiles #1008

### PR DESCRIPTION
**Pull request by devnewton**
I have implemented the tile sheet with different sized tiles support with:

- only two (optional) addition to the xml/json format : top and left attributes for image tag ;
- a little GUI change: a new image rect property.

The goal is to use manually drawn spritesheets through tile collection without breaking Tiled normal workflow/code.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<tileset version="1.8" tiledversion="1.8.4" name="plop" tilewidth="100" tileheight="200" tilecount="3" columns="0">
 <grid orientation="orthogonal" width="1" height="1"/>
 <tile id="0">
  <image width="100" height="200" top="10" left="20" source="squirrel.png"/>
 </tile>
 <tile id="1">
  <image width="100" height="100" top="0" left="200" source="squirrel.png"/>
 </tile>
 <tile id="2">
  <image width="95" height="48" top="50" left="878" source="squirrel.png"/>
 </tile>
</tileset>
```

image

image

image

### CODE REVIEW

**Review by bjorn — CHANGES_REQUESTED** [review]
Thanks for getting this feature started! As it is, I think there are some problems that we'll need to address, see also my inline comments.

The main issue is as I mentioned at https://github.com/mapeditor/tiled/issues/1008#issuecomment-838154607, we're looking for a solution to this that also works for tileset images. If we just add `top` and `left` attributes to the `image` element, then there is no way to cover the other use-case. Besides that, re-using the `width` and `height` attributes is a compatibility issue and leads to unexpected behavior when images are resized.

I've just proposed an alternative storage at https://github.com/mapeditor/tiled/issues/2863#issuecomment-1117500495, and I think we could adjust this PR to use that format. I'd welcome your thoughts on this!

File: src/libtiled/maptovariantconverter.cpp [hunk 1]
```diff
@@ -287,6 +287,11 @@ QVariant MapToVariantConverter::toVariant(const Tileset &tileset,
                 tileVariant[QStringLiteral("imagewidth")] = tileSize.width();
                 tileVariant[QStringLiteral("imageheight")] = tileSize.height();
             }
+            const QPoint topLeft = tile->imageRect().topLeft();
+            if (!topLeft.isNull()) {
+                tileVariant[QStringLiteral("imagetop")] = topLeft.y();
+                tileVariant[QStringLiteral("imageleft")] = topLeft.x();
+            }
```
[the remark this answers is not in the document]
Comment (devnewton) [reply src/libtiled/maptovariantconverter.cpp:297]: I've removed the top/left stuff.

File: src/libtiled/imagereference.cpp [hunk 2]
```diff
@@ -32,14 +32,15 @@ bool ImageReference::hasImage() const
...
+        pixmap = ImageCache::loadPixmap(source.toLocalFile());
     else if (source.scheme() == QLatin1String("qrc"))
-        return ImageCache::loadPixmap(QLatin1Char(':') + source.path());
+        pixmap = ImageCache::loadPixmap(QLatin1Char(':') + source.path());
     else if (!data.isEmpty())
-        return QPixmap::fromImage(QImage::fromData(data, format));
+        pixmap = QPixmap::fromImage(QImage::fromData(data, format));
 
-    return QPixmap();
+    return pixmap.copy(QRect(topLeft, size));
```
[the remark this answers is not in the document]
Comment (devnewton) [reply src/libtiled/imagereference.cpp]: I reverted this pixmap copy.

**Comment by devnewton** [comment]
I've made the suggested change to use x,y,width,height properties at tile level: https://github.com/mapeditor/tiled/pull/3339/commits/783284cce2d42cc4c7b3932f61b03c88a58dc39a

File: src/tiled/tileanimationeditor.cpp [hunk 3]
```diff
@@ -428,6 +431,19 @@ void TileAnimationEditor::framesEdited()
...
 }
 
+void TileAnimationEditor::tilesetChanged()
+{
+    auto *tilesetDocument = static_cast<TilesetDocument*>(sender());
+    auto *tilesetView = mUi->tilesetView;
+    auto *model = tilesetView->tilesetModel();
+
+    if (tilesetDocument == mTilesetDocument)
+        setTile(nullptr);        // It may be gone
```
[the remark this answers is not in the document]
Comment (bjorn) [reply src/tiled/tileanimationeditor.cpp:441]: @devnewton Can you shed some light on why you added this?

**Review by bjorn — APPROVED** [review]
Sometimes it's hard to add even simple features, given the amount of existing code that it will affect. :-)

I think the feature should work reliably now, but I may of course have missed some cases as well. Could you please also have a final look and do some testing with your use-cases, to see whether the change is fine in its current state?

Personally, I'd like to continue this change to cover also tiles from a tileset image (see also #2863). If we store the sub-rectangle on each tile, we can entirely avoid the need to even cut out each tile's image from the tileset image (apart from where it is used as an icon). But I guess it's better to continue that on another pull request.

**Comment by devnewton** [comment]
I have made some tests and it works well. UX/UI could be better, but this will be for other pull requests :-)

Here is the test map I used, feel free to add it as Tiled sample. It show background object layers, a tile layer, and an animated object with various tile sizes.

[forest.zip](https://github.com/mapeditor/tiled/files/8708403/forest.zip)
)

**Comment by bjorn** [comment]
> Here is the test map I used, feel free to add it as Tiled sample. It show background object layers, a tile layer, and an animated object with various tile sizes.

Thanks! I've just committed this example then.

**Pull request merged**